invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,198 @@
+type invoiceExpiry struct {
+	PaymentHash lntypes.Hash
+	Expiry      time.Time
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
+func (e invoiceExpiry) Less(other queue.PriorityQueueItem) bool {
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
+// invoices. Upon start InvoiceExpiryWatcher will retrieve all pending (not yet
+// settled or canceled) invoices invoices to its watcing queue. When a new
+// invoice is added to the InvoiceRegistry, it'll be forarded to the
+// InvoiceExpiryWatcher and will end up in the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice().
+type InvoiceExpiryWatcher struct {
+	sync.Mutex
+	started bool
+	wg      sync.WaitGroup
+
+	// clock is the clock implementation that InvoiceExpiryWatcher uses.
+	// It is useful for testing.
+	clock clock.Clock
+
+	// quit signals InvoiceExpiryWatcher to stop.
+	quit chan interface{}
+
+	// cancelInvoice is a template method that cancels an expired invoice.
+	cancelInvoice func(lntypes.Hash) error
+
+	// expiryQueue holds invoiceExpiry items and is used to find the next
+	// invoice to expire.
+	expiryQueue queue.PriorityQueue
+
+	// newInvoices channel is used to wake up the main loop when a new invoices
+	// is added.
+	newInvoices chan *invoiceExpiry
+}
+
+// NewInvoiceExpiryWatcher creates a new InvoiceExpiryWatcher instance.
+func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+	return &InvoiceExpiryWatcher{
+		clock:       clock,
+		quit:        make(chan interface{}),
+		newInvoices: make(chan *invoiceExpiry),
+	}
+}
+
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher queue. This is useful to
+// prepopulate InvoiceExpiryWatcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) error {
+	pendingOnly := true
+	pendingInvoices, err := cdb.FetchAllInvoicesWithPaymentHash(pendingOnly)
+	if err != nil && err != channeldb.ErrNoInvoicesCreated {
+		log.Errorf(
+			"Error while prefetching active invoices from the database: %v", err,
+		)
+		return err
+	}
+
+	for k := 0; k < len(pendingInvoices); k++ {
+		ew.AddInvoice(pendingInvoices[k].PaymentHash, &pendingInvoices[k].Invoice)
+	}
+
+	return nil
+}
+
+// Start starts the the subscription handler and the main loop. Start() will
+// return with error if InvoiceExpiryWatcher is already started. Start()
+// expects a cancellation function passed that will be use to cancel expired
+// invoices by their payment hash.
+func (ew *InvoiceExpiryWatcher) Start(
+	cancelInvoice func(lntypes.Hash) error) error {
+	ew.Lock()
+	defer ew.Unlock()
+
+	if ew.started {
```

Comment:
A pattern that we use elsewhere is an atomic add: `if atomic.AddInt32(&n.started, 1) != 1 {`

Then the lock isn't needed.

Reply:
I understand that, but to me that pattern fails when for some reason (early stopping) `Stop()` is called during `Start()`. By locking `Stop()` will wait and do the stopping, with compare exchange it'll simply return.